pkg/apk/impl/repo.go

```diff
@@ -0,0 +1,443 @@
+		pkgProvidesMap  = map[string]string{}
+		dependencies    = make([]*repository.RepositoryPackage, 0, 20)
+		dependenciesMap = map[string]bool{}
+	)
+	// create a map of every package to its RepositoryPackage
+	for _, index := range indexes {
+		for _, pkg := range index.Packages() {
+			if _, ok := pkgNameMap[pkg.Name]; !ok {
+				pkgNameMap[pkg.Name] = pkg
+			}
+		}
+	}
+	// create a map of every provided file to its package
+	for _, pkg := range pkgNameMap {
+		for _, provide := range pkg.Provides {
+			name, _, compare := resolvePackageNameVersion(provide)
+			if _, ok := pkgProvidesMap[name]; !ok {
+				pkgProvidesMap[name] = pkg.Name
+			}
+			if compare != versionNone {
+				if _, ok := pkgProvidesMap[provide]; !ok {
+					pkgProvidesMap[provide] = pkg.Name
+				}
+			}
+		}
+	}
+	for _, pkgName := range packages {
+		// this package might be pinned to a version
+		name, version, compare := resolvePackageNameVersion(pkgName)
+		pkg, ok := pkgNameMap[name]
+		if !ok {
+			return nil, nil, fmt.Errorf("could not find package %s in indexes", pkgName)
+		}
+		if err := compareVersion(pkg.Version, version, compare); err != nil {
+			return nil, nil, fmt.Errorf("mismatched version for package %s in indexes: %w", pkgName, err)
+		}
+		deps, confs, err := getPackageDependencies(name, pkgNameMap, pkgProvidesMap)
+		if err != nil {
+			return nil, nil, err
+		}
+		dependencies = append(dependencies, pkg)
+		dependencies = append(dependencies, deps...)
+		conflicts = append(conflicts, confs...)
+	}
+	// The entire list of dependencies now is here; need to rationalize it
+	// to remove duplicates. It is sorted by lowest dependency last, so we should be able to
+	// go through it in reverse, dropping newer ones as they show up.
+	for i := len(dependencies) - 1; i >= 0; i-- {
+		dep := dependencies[i]
+		// now we have the new dependency, check if it's already in the list
+		if _, ok := dependenciesMap[dep.Name]; ok {
+			continue
+		}
+		toInstall = append(toInstall, dep)
+		dependenciesMap[dep.Name] = true
+	}
+
+	return toInstall, conflicts, nil
+}
+
+// getPackageDependencies get all of the dependencies for a single package based on the
+// indexes.
+func getPackageDependencies(pkg string, pkgNameMap map[string]*repository.RepositoryPackage, pkgProvidesMap map[string]string) (dependencies []*repository.RepositoryPackage, conflicts []string, err error) {
+	ref, ok := pkgNameMap[pkg]
+	if !ok {
+		return nil, nil, fmt.Errorf("package %s not found in indexes", pkg)
+	}
+
+	// each dependency has several possibilities:
+	// - /bin/sh   - this means, "I need package busybox"
+	// - name      - this means, "I need package 'name'"
+	// - so:<lib>  - this means, "find the package that provides so:<lib>"
+	// - cmd:<cmd> - this means, "find the package that provides cmd:<cmd>"
+	for _, dep := range ref.Dependencies {
+		var newDep string
+		switch {
+		case dep == "/bin/sh":
+			newDep = "busybox"
```

Comment:
`/bin/sh` is a virtual, which is presently provided by `busybox`, but might be provided by something else in the future.  I think we should not have special handling for it.

The upstream `apk` solver algorithm looks for a direct package which matches `apk_name`, and then falls back to looking for packages which *provide* the `apk_name`.  This would remove the need for special casing `so:` and `cmd:` in your algorithm as well (and there other namespaces like `py:`, `ruby:`, `pc:` in some apk distributions).

Reply:
It is clear you have dug more deeply into how it does it than I have (but the gap has narrowed, if only a bit).

My understanding was that the algorithm is:

1. dep is `!<something>` (special case)
1. dep is package name
1. dep is a special case (i.e. `/bin/sh`)
1. dep is something provided by a package, when it has the `so:` or `cmd:` cases

You are saying (I think):

1. dep is `!<something>` (special case)
1. dep is package name
1. dep is any package that _provides_ that dependency

The last case covers package alternate names, but also `/bin/sh` (since it is included in the `p:` of busybox), but also the `so:` and `cmd:` etc.

Is that correct?

Reply:
The `!` just means "conflicts with <thing>".  Otherwise your analysis is correct.

Reply:
@kaniini I cleaned it up. The logic now is essentially what is above.

With one variant.

I had to add explicit support for the virtual `/bin/sh`. Without it, anything that provides `/bin/sh` could satisfy it, which leads to strange loops. For example:

1. I already have `busybox`, which provides `/bin/sh`
1. I want to install `dash-binsh`, which also provides `/bin/sh`
2. That depends on `dash`
3. That depends on `/bin/sh`

So do I satisfy it with `busybox` or `dash-binsh`. We know that it means the baseline `/bin/sh`, or perhaps anything that provides it. The dependency is created because `dash` has a pre-install script that requires `/bin/sh`, but how would this resolve it?

I guess it is possible to say, "as long as something has provided it, we are good"?

An alternative scenario is where I install something that depends on `/bin/sh`, e.g. `dhcp`. When it asks for `/bin/sh`, if there is none installed yet, do I install `busybox`? `dash`? `yash`?